src/assets/TokenDetectionController.test.ts

```diff
@@ -348,119 +315,204 @@ describe('TokenDetectionController', () => {
       .stub(tokensController, '_instantiateNewEthersProvider')
       .callsFake(() => null);
-    const firstNetworkType = 'rinkeby';
-    const secondNetworkType = 'mainnet';
-    const firstAddress = '0x123';
-    const secondAddress = '0x321';
-    const detectTokens = sinon.stub(tokenDetection, 'detectTokens');
-    preferences.update({ selectedAddress: secondAddress });
-    preferences.update({ selectedAddress: secondAddress });
-    expect(preferences.state.selectedAddress).toStrictEqual(secondAddress);
-    expect(detectTokens.calledTwice).toBe(false);
-    preferences.update({ selectedAddress: firstAddress });
-    expect(preferences.state.selectedAddress).toStrictEqual(firstAddress);
+
+    tokenDetection.configure({
+      selectedAddress: '0x1',
+    });
+    const detectedTokensMock = sinon
+      .stub(tokenDetection, 'detectTokens')
+      .callsFake(() => Promise.resolve());
+
     network.update({
       provider: {
-        type: secondNetworkType,
-        chainId: NetworksChainId[secondNetworkType],
+        type: 'rpc',
+        chainId: '56',
       },
     });
-    expect(network.state.provider.type).toStrictEqual(secondNetworkType);
+
+    await tokenDetection.start();
+    expect(detectedTokensMock.callCount).toBe(2);
+
+    tokenDetection.stop();
     network.update({
       provider: {
-        type: firstNetworkType,
-        chainId: NetworksChainId[firstNetworkType],
+        type: 'rpc',
+        chainId: '137',
       },
     });
-    expect(network.state.provider.type).toStrictEqual(firstNetworkType);
-    tokensController.update({ tokens: TOKENS });
-    expect(tokenDetection.config.tokens).toStrictEqual(TOKENS);
+
+    // Detected tokens will still be called once more but will not actually detect tokens
+    expect(detectedTokensMock.callCount).toBe(3);
+  });
+
+  it('should not call detectedTokens from onTokenListStateChange if tokenList is empty', async () => {
+    const stub = sinon.stub();
+    const detectTokensMock = sinon.stub(tokenDetection, 'detectTokens');
+    let tokenListStateChangeListener: (state: any) => void;
+    const onTokenListStateChange = sinon.stub().callsFake((listener) => {
+      tokenListStateChangeListener = listener;
+    });
+    tokenDetection = new TokenDetectionController(
+      {
+        onTokenListStateChange,
+        onPreferencesStateChange: stub,
+        onNetworkStateChange: stub,
+        getBalancesInSingleCall: stub,
+        addDetectedTokens: stub,
+        getTokensState: stub,
+        getTokenListState: stub,
+        getNetworkState: () => network.state,
+        getPreferencesState: () => preferences.state,
+      },
+      {
+        disabled: false,
+        isDetectionEnabledForNetwork: true,
+        isDetectionEnabledFromPreferences: true,
+      },
+    );
+
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    await tokenListStateChangeListener!({ tokenList: {} });
+
+    expect(detectTokensMock.called).toBe(false);
```

Comment:
Nit: Could this assertion look at `getBalancesInSingleCall` instead, or some other "external" signal for whether tokens are being detected or not?

That the token list listener calls `detectTokens` is just an implementation detail. Expressing this assertion without referring to controller internals would make this test less likely to break when the controller is refactored.

This applies to the next two tests as well

Reply:
Updated